lib/phoenix/presence.ex

```diff
@@ -343,29 +343,203 @@ defmodule Phoenix.Presence do
       pubsub_server =
         opts[:pubsub_server] || raise "use Phoenix.Presence expects :pubsub_server to be given"
 
-      Phoenix.Tracker.start_link(__MODULE__, {module, task_supervisor, pubsub_server}, opts)
+      presence_client = opts[:presence_client]
+
+      Phoenix.Tracker.start_link(
+        __MODULE__,
+        {module, task_supervisor, pubsub_server, presence_client},
+        opts
+      )
     end
 
     def init(state) do
       {:ok, state}
     end
 
     def handle_diff(diff, state) do
-      {module, task_supervisor, pubsub_server} = state
+      {module, task_supervisor, pubsub_server, presence_client} = state
 
       Task.Supervisor.start_child(task_supervisor, fn ->
         for {topic, {joins, leaves}} <- diff do
-          Phoenix.Channel.Server.local_broadcast(pubsub_server, topic, "presence_diff", %{
+          presence_diff = %{
             joins: module.fetch(topic, Phoenix.Presence.group(joins)),
             leaves: module.fetch(topic, Phoenix.Presence.group(leaves))
-          })
+          }
+
+          if presence_client do
+            Phoenix.Presence.Client.handle_diff(topic, presence_diff)
+          end
+
+          Phoenix.Channel.Server.local_broadcast(
+            pubsub_server,
+            topic,
+            "presence_diff",
+            presence_diff
+          )
         end
       end)
 
       {:ok, state}
     end
   end
 
+  defmodule Client do
+    use GenServer
+    require Logger
+
+    @callback init(state :: term) :: {:ok, new_state :: term}
+    @callback handle_join(topic :: String.t(), key :: String.t(), meta :: [map()], state :: term) ::
```

Comment:
as discussed, let's consolidate `handle_join` and `handle_leave` into `handle_metas`, which can look something like this in user land:

```elixir
      def handle_metas(topic, %{joins: _, leaves: _} = diff, presences, state) do
        for {user_id, meta} <- diff.joins do
          presence = Map.fetch!(presences, user_id)
          local_broadcast(topic, {__MODULE__, %{user_joined: presence}})
        end

        for leave <- diff.leaves do
          local_broadcast(topic, {__MODULE__, %{user_left: presence}})
        end
        {:ok, state}
      end
```